The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4087940492/jobs/7049045954) [(plain)](https://github.com/rust-lang-ci/rust/commit/90c1e6c689773c8d55031328f01ca276da9d157d/checks/11100172097/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
drwxrwxrwt 14 root root 4.0K Feb  3 22:01 ..
-rw-r--r--  1 gha  gha  125K Feb  3 22:01 cpu-aarch64-gnu.csv
-rw-r--r--  1 gha  gha   99K Feb  3 22:01 metrics-aarch64-gnu.json

Attempting with retry: aws s3 cp --storage-class INTELLIGENT_TIERING --no-progress --recursive --acl public-read /tmp/tmp.GVaKeMDjnt s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/cpu-aarch64-gnu.csv to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv"
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/metrics-aarch64-gnu.json to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json"
Command failed. Attempt 2/5:
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/cpu-aarch64-gnu.csv to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv"
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/metrics-aarch64-gnu.json to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json"
Command failed. Attempt 3/5:
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/cpu-aarch64-gnu.csv to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.us-west-1.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv"
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/metrics-aarch64-gnu.json to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.us-west-1.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json"
Command failed. Attempt 4/5:
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/cpu-aarch64-gnu.csv to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.us-west-1.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv"
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/metrics-aarch64-gnu.json to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.us-west-1.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json"
Command failed. Attempt 5/5:
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/cpu-aarch64-gnu.csv to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.us-west-1.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/cpu-aarch64-gnu.csv"
upload failed: ../../../../tmp/tmp.GVaKeMDjnt/metrics-aarch64-gnu.json to s3://rust-lang-ci2/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json Could not connect to the endpoint URL: "https://rust-lang-ci2.s3.us-west-1.amazonaws.com/rustc-builds/90c1e6c689773c8d55031328f01ca276da9d157d/metrics-aarch64-gnu.json"
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_infer/src/infer/nll_relate/mod.rs
+++ b/compiler/rustc_infer/src/infer/nll_relate/mod.rs
@@ -817,12 +817,13 @@ impl<'tcx, D> ConstEquateRelation<'tcx> for TypeRelating<'_, 'tcx, D>
 where
     D: TypeRelatingDelegate<'tcx>,
 {
-    fn const_equate_obligation(&mut self, _a: ty::Const<'tcx>, _b: ty::Const<'tcx>) {
-        // We don't have to worry about the equality of consts during borrow checking
-        // as consts always have a static lifetime.
-        // FIXME(oli-obk): is this really true? We can at least have HKL and with
-        // inline consts we may have further lifetimes that may be unsound to treat as
-        // 'static.
+    fn const_equate_obligation(&mut self, a: ty::Const<'tcx>, b: ty::Const<'tcx>) {
+        self.delegate.register_obligations(vec![Obligation::new(
+            self.tcx(),
+            ObligationCause::dummy(),
+            self.param_env(),
+            ty::Binder::dummy(ty::PredicateKind::ConstEquate(a, b)),
+        )]);
     }
 }
 

--- a/compiler/rustc_middle/src/ty/relate.rs
+++ b/compiler/rustc_middle/src/ty/relate.rs
@@ -629,6 +629,8 @@ pub fn super_relate_consts<'tcx, R: TypeRelation<'tcx>>(
         b = tcx.expand_abstract_consts(b);
     }
 
+    debug!("{}.super_relate_consts(normed_a = {:?}, normed_b = {:?})", relation.tag(), a, b);
+
     // Currently, the values that can be unified are primitive types,
     // and those that derive both `PartialEq` and `Eq`, corresponding
     // to structural-match types.
@@ -665,30 +667,28 @@ pub fn super_relate_consts<'tcx, R: TypeRelation<'tcx>>(
 
             // FIXME(generic_const_exprs): is it possible to relate two consts which are not identical
             // exprs? Should we care about that?
+            // FIXME(generic_const_exprs): relating the `ty()`s is a little weird since it is supposed to
+            // ICE If they mismatch. Unfortunately `ConstKind::Expr` is a little special and can be thought
+            // of as being generic over the argument types, however this is implicit so these types don't get
+            // related when we relate the substs of the item this const arg is for.
             let expr = match (ae, be) {
-                (Expr::Binop(a_op, al, ar), Expr::Binop(b_op, bl, br))
-                    if a_op == b_op && al.ty() == bl.ty() && ar.ty() == br.ty() =>
-                {
+                (Expr::Binop(a_op, al, ar), Expr::Binop(b_op, bl, br)) if a_op == b_op => {
+                    r.relate(al.ty(), bl.ty())?;
+                    r.relate(ar.ty(), br.ty())?;
                     Expr::Binop(a_op, r.consts(al, bl)?, r.consts(ar, br)?)
                 }
-                (Expr::UnOp(a_op, av), Expr::UnOp(b_op, bv))
-                    if a_op == b_op && av.ty() == bv.ty() =>
-                {
+                (Expr::UnOp(a_op, av), Expr::UnOp(b_op, bv)) if a_op == b_op => {
+                    r.relate(av.ty(), bv.ty())?;
                     Expr::UnOp(a_op, r.consts(av, bv)?)
                 }
-                (Expr::Cast(ak, av, at), Expr::Cast(bk, bv, bt))
-                    if ak == bk && av.ty() == bv.ty() =>
-                {
+                (Expr::Cast(ak, av, at), Expr::Cast(bk, bv, bt)) if ak == bk => {
+                    r.relate(av.ty(), bv.ty())?;
                     Expr::Cast(ak, r.consts(av, bv)?, r.tys(at, bt)?)
                 }
                 (Expr::FunctionCall(af, aa), Expr::FunctionCall(bf, ba))
-                    if aa.len() == ba.len()
-                        && af.ty() == bf.ty()
-                        && aa
-                            .iter()
-                            .zip(ba.iter())
-                            .all(|(a_arg, b_arg)| a_arg.ty() == b_arg.ty()) =>
+                    if aa.len() == ba.len() =>
                 {
+                    r.relate(af.ty(), bf.ty())?;
                     let func = r.consts(af, bf)?;
                     let mut related_args = Vec::with_capacity(aa.len());
                     for (a_arg, b_arg) in aa.iter().zip(ba.iter()) {

--- a/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_binop_arg_tys.rs
+++ b/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_binop_arg_tys.rs
@@ -1,0 +1,18 @@
+// checks that when we relate a `Expr::Binop` we also relate the types of the
+// const arguments.
+#![feature(generic_const_exprs)]
+#![allow(incomplete_features)]
+
+struct Bar<const B: bool>;
+
+const fn make_generic(_: usize, a: bool) -> bool {
+    a
+}
+
+fn foo<const N: usize>() -> Bar<{ make_generic(N, true == false) }> {
+    Bar::<{ make_generic(N, 1_u8 == 0_u8) }>
+    //~^ error: mismatched types
+    //~| error: unconstrained generic constant
+}
+
+fn main() {}

--- a/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_binop_arg_tys.stderr
+++ b/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_binop_arg_tys.stderr
@@ -1,0 +1,20 @@
+error[E0308]: mismatched types
+  --> $DIR/relate_binop_arg_tys.rs:13:5
+   |
+LL |     Bar::<{ make_generic(N, 1_u8 == 0_u8) }>
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `{ make_generic(N, true == false) }`, found `{ make_generic(N, 1_u8 == 0_u8) }`
+   |
+   = note: expected constant `{ make_generic(N, true == false) }`
+              found constant `{ make_generic(N, 1_u8 == 0_u8) }`
+
+error: unconstrained generic constant
+  --> $DIR/relate_binop_arg_tys.rs:13:11
+   |
+LL |     Bar::<{ make_generic(N, 1_u8 == 0_u8) }>
+   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = help: try adding a `where` bound using this expression: `where [(); { make_generic(N, 1_u8 == 0_u8) }]:`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_cast_arg_ty.rs
+++ b/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_cast_arg_ty.rs
@@ -1,0 +1,12 @@
+// checks that when we relate a `Expr::Cast` we also relate the type of the
+// const argument.
+#![feature(generic_const_exprs)]
+#![allow(incomplete_features)]
+
+fn foo<const N: usize>() -> [(); (true as usize) + N] {
+    [(); (1_u8 as usize) + N]
+    //~^ error: mismatched types
+    //~| error: unconstrained generic constant
+}
+
+fn main() {}

--- a/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_cast_arg_ty.stderr
+++ b/tests/ui/const-generics/generic_const_exprs/const_kind_expr/relate_cast_arg_ty.stderr
@@ -1,0 +1,20 @@
+error[E0308]: mismatched types
+  --> $DIR/relate_cast_arg_ty.rs:7:5
+   |
+LL |     [(); (1_u8 as usize) + N]
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `(true as usize) + N`, found `(1_u8 as usize) + N`
+   |
+   = note: expected constant `(true as usize) + N`
+              found constant `(1_u8 as usize) + N`
+
+error: unconstrained generic constant
+  --> $DIR/relate_cast_arg_ty.rs:7:10
+   |
+LL |     [(); (1_u8 as usize) + N]
+   |          ^^^^^^^^^^^^^^^^^^^
+   |
+   = help: try adding a `where` bound using this expression: `where [(); (1_u8 as usize) + N]:`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

